### ISSUE TITLE
add async and idempotency support to azurerm_managed_disk

### ISSUE THREAD

**Issue by rbreitendev** [ansible/ansible_collections_azure#847]
##### SUMMARY

When attempting to attach a large number of disks to many (100+) systems at scale via the Azure `azurerm_managed_disk` module, `async` support should allow the ability to independently create and attach multiple disks across a large number of Azure virtual machines. This supported configuration exists with other competitors and is in use in many of our production environments, allowing for attachment of a dozen or more disks across hundreds of systems.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

azurerm_managed_disk

##### ADDITIONAL INFORMATION

Both Azure Commercial and AzureUSGovernment are affected cloud providers. This works inside other disparate cloud providers for both commercial and higher security compliance requirements without issue. The below code snippet gets us around it temporarily leveraging the REST API but this is unsustainable long-term.

```yaml
name: Attach disks to Azure VM via Rest API
when: disk_create | bool and (data_disk_json | length > 0)
delegate_to: "{{ task_delegation | default(omit, true) }}"
azure_rm_resource:
resource_group: "{{ azure_resource_group }}"
provider: compute
resource_type: virtualMachines
resource_name: "{{ azure_vm_name }}"
api_version: "2021-11-01"
body: {
"location": "{{ azure_location }}",
"properties": {
    "storageProfile": {
        "dataDisks": "{{ data_disk_json }}"
    }
}
}
```

**Comment by Fred-sun** [comment]
@rbreitendev Thank you for reporting this problem. The idempotent problem is clear and easy to solve. But what do you mean by async? Could you describe it in detail? Thank you very much!

**Comment by Fred-sun** [comment]
Add multiple data disks to a VM at the same time, or add a data disk to multiple VMS at the same time? Thank you very much!

**Comment by sscheib-rh** [comment]
@Fred-sun  It means to use the Ansible keyword "async" [1] in a loop while creating and attaching multiple disks to multiple VMs, which greatly speeds up the execution time.


[1]: https://docs.ansible.com/ansible/latest/user_guide/playbooks_async.html

**Comment by Fred-sun** [comment]
@sscheib-rh This async is to limit the execution time of a task, that is, the time interval to the next task. If you want to mount multiple disks to multiple VMS at the same time, you can configure multi-task by code. Thank you very much!

**Comment by rbreitendev** [comment]
@Fred-sun essentially, I want to be able to attach 5+ volumes per VM across 100 VM's at a time. In other providers, we are able to perform this action leveraging `async` so that the VM's have no issue with attaching those disks and initializing/formatting them for us within the VM itself. We were previously told this was impossible within Azure and to leverage the rest API instead. If this exists within the API, why is the functionality not within the Ansible module as well?

**Comment by Fred-sun** [comment]
@rbreitendev The azure_rm_manageddisk module calls the API to create or attach an existing disk to a VM. If you mount it to multiple VMS, it still requires multitasking code!

**Comment by sscheib-rh** [comment]
@Fred-sun It would be sufficient if azure_rm_manageddisk would have the possibility to create and attach multiple disks simultaneously to one VM. Is it possible to provide such functionality?

**Comment by Fred-sun** [comment]
@sscheib-rh Whether the disk already exists or whether the disk is created and attached to the VM. Thank you very much!

**Comment by sscheib-rh** [comment]
@Fred-sun In this particular case the disk**s** would not exist, which means they would need to be created and attached.

**Comment by Fred-sun** [comment]
@sscheib-rh I see! I will check and reply asap! Thank you very much!

**Comment by Fred-sun** [comment]
@sscheib-rh You can manage and configure disks using the "azure_rm_manageddisk" module, and then mount multiple data disks to the VM using the "azure_rm_virtualmachine" module. Thank you very much!

**Comment by Fred-sun** [comment]
@rbreitendev @sscheib-rh Now we have the following two solutions, can they meet your needs?

First: With the new func in Pull Request (https://github.com/ansible-collections/azure/pull/867), you can create or manage disks and mount them to multiple VMS, as shown in the following use case:
```e.g.
    - name: Mount the managed disk to multiple VMs
      azure_rm_manageddisk:
        resource_group: myResourceGroup
        name: freddisk04
        max_shares: 4
        disk_size_gb: 1024
        storage_account_type: Premium_LRS
        managed_by_extended:
          - resource_group: myResourceGroup01
            name: testVM01
          - resource_group: myResourceGroup02
            name: testVM02
        zone: 1
```
Second: You can use azure_rm_virtualmachine to create multiple disks and mount them to a specified VM, as shown in the following case:
```e.g.
    - name: Create Resource group
      azure_rm_resourcegroup:
        name: "{{ resource_group }}"
        location: eastus

    - name: Create virtual network
      azure_rm_virtualnetwork:
        resource_group: "{{ resource_group }}"
        name: "{{ network_name }}"
        address_prefixes: 10.42.0.0/24

    - name: Create subnet
      azure_rm_subnet:
        resource_group: "{{ resource_group }}"
        name: "{{ subnet_name }}"
        address_prefix: 10.42.0.0/28
        virtual_network: "{{ network_name }}"

    - name: Create a VM multiple data managed disks
      azure_rm_virtualmachine:
        resource_group: "{{ resource_group }}"
        name: "{{ vm_name }}"
        vm_size: Standard_D4s_v3
        managed_disk_type: Standard_LRS
        admin_username: azureuser
        admin_password: Password@0329
        ssh_public_keys:
          - path: /home/azureuser/.ssh/authorized_keys
            key_data: "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQDBdxN6iGt/BEv2FCPzJU2pbyKLNIiFQKbT7p/+Meh2Z/2/JcfRg1O/TuQoQSxFanxz/62iy1J5cOouE8DJa1EFvbq1yC2FafHE7rIVoQdgO9SDkIKOIapI0/EB1A9fh6Qkem5VM19zuyParA4Cc3s2eonzBX2zXza3SfpK97OJIMtMqy97eQzrhpx7XyrVyJsor3rlMAcq5EhG6I7O9rWN42vsKf5z8/0iwR5dbClVfEm0O3qcJxCZE748PxCMWnimjoP7RnCSSXd6KVsEDSx7f8e/Gvul6Yz46HzmhdFBfzPZxz+iOP77k4S1LwvE9EOE8o97MUTIncC+h43N2FLUgIraG/goKb2l1yRY+8yOzu+Hy/C6+Q88/eYgDFprFcUSlnxRpNxpqjlE/Ox6wcny6YXIq6BM5dAUfRMccpNH/k4tFWo6wBoLUDw1KEAc3Fv2/PmbNUFcSYTwXWHJbSHspDJNWOCQD1qcZZGX1bDZr3+RlEKUkScOfTpMFDUsCjM= xiuxi.sun@qq.com"
        image:
          offer: UbuntuServer
          publisher: Canonical
          sku: 18.04-LTS
          version: latest
        data_disks:
          - lun: 0
            disk_size_gb: 1024
            managed_disk_type: Premium_LRS
          - lun: 1
            disk_size_gb: 128
            managed_disk_type: Premium_LRS
          - lun: 2
            disk_size_gb: 1024
            managed_disk_type: Premium_LRS

    - name: Get VM info
      azure_rm_virtualmachine_info:
        resource_group: "{{ resource_group }}"
        name: "{{ vm_name }}"
      register: output

    - assert:
         that:
           - output.vms[0].data_disks | length == 3

**Comment by archoversight** [comment]
Neither suggestion meets our requirements.

We have 100 VM's that we run ansible across, each of those VM's needs 5 or more disk attached.

It is a 1:1 relation for disk to VM. We do not have the same disk mounted to multiple VM's.

So that is 500 disk that needs to be created and attached. The VM is already running and is managed using terraform, we will not be switching to ansible for managing that VM.

Without async support it looks like this:

1. VM A attaches disk 1
2. VM B attaches disk 1
3. VM C attaches disk 1
4. VM A attaches disk 2
5. VM B attaches disk 2
6. VM C attaches disk 2
7. ...
8. VM A attaches disk 5
9. VM B attaches disk 5
10. VM C attaches disk 5

This all happens one by one, Ansible can not and won't speed this up as its done synchronously unless you use the async keyword in the loop so that it continues down the loop. As you can imagine with 100+ machines each needing X amount of disks, the time it takes to complete is incredibly long.

**Comment by Fred-sun** [comment]
@archoversight Thanks for sharing your thoughts, this needs to change the current logic, add multitasking support. We'll discuss it and decide. Thank you very much!

**Comment by Fred-sun** [comment]
@archoversight I have added multitasking support in PR #909. When mounting VM, you can attach disks to multiple VMS at the same time. Does this satisfy your needs?

**Comment by xuzhang3** [comment]
> Neither suggestion meets our requirements.
> 
> We have 100 VM's that we run ansible across, each of those VM's needs 5 or more disk attached.
> 
> It is a 1:1 relation for disk to VM. We do not have the same disk mounted to multiple VM's.
> 
> So that is 500 disk that needs to be created and attached. The VM is already running and is managed using terraform, we will not be switching to ansible for managing that VM.
> 
> Without async support it looks like this:
> 
> 1. VM A attaches disk 1
> 2. VM B attaches disk 1
> 3. VM C attaches disk 1
> 4. VM A attaches disk 2
> 5. VM B attaches disk 2
> 6. VM C attaches disk 2
> 7. ...
> 8. VM A attaches disk 5
> 9. VM B attaches disk 5
> 10. VM C attaches disk 5
> 
> This all happens one by one, Ansible can not and won't speed this up as its done synchronously unless you use the async keyword in the loop so that it continues down the loop. As you can imagine with 100+ machines each needing X amount of disks, the time it takes to complete is incredibly long.

@archoversight @rbreitendev #909 will async attach the disk to mulit-VMs

**Comment by archoversight** [comment]
@xuzhang3 looking at the code it only modifies the path that says it is attaching the same disk to multiple VM's (See comment here: https://github.com/ansible-collections/azure/pull/909/files#diff-eb88fb554c994eddda6e44d62f486d9e57d26ff4f320979a43cb13b8a62d7600L456). We have `M` disks, which get attached to N machines. Where `N` is for example `M / 5` (we attach 5 disks to 1 single machine). Each disk has a 1:1 relationship with the VM it is attached to.

We do not attach `1` disk to `Y` machines where `Y > 1`.

I am sorry I did not make that more clear in my example.

I did say:

> It is a 1:1 relation for disk to VM. We do not have the same disk mounted to multiple VM's.

Updated example:

- VM A attaches disk 1A
- VM B attaches disk 1B
- VM C attaches disk 1C
- VM A attaches disk 2A
- VM B attaches disk 2B
- VM C attaches disk 2C
- ...
- VM A attaches disk 5A
- VM B attaches disk 5B
- VM C attaches disk 5C

**Comment by archoversight** [comment]
Here's some sample code that showcases how this works with EC2:

```
    - name: Create and tag the additonal volumes on AWS using ec2_vol
      ec2_vol:
        device_name: '{{ item }}'
        instance: '{{ ansible_ec2_instance_id }}'
        volume_type: '{{ disk_preset_list[index].value.type | default("gp2") }}'
        throughput: '{{disk_preset_list[index].value.throughput | default(omit, true) }}'
        iops: '{{disk_preset_list[index].value.iops | default(omit, true) }}' # Max 500:1 disk size
        encrypted: '{{ disk_encryption }}'
        kms_key_id: '{{ disk_preset_list.value.disk_encryption_key | default(disk_encryption_key_default) }}'
        volume_size: '{{ disk_preset_list[index].value.size|int }}'
        zone: '{{ ansible_ec2_placement_availability_zone }}'
        region: '{{ ansible_ec2_placement_region }}'
        tags:
          Name: '{{ volume_name_randomization_string }}_{{ disk_preset_list[index].key }}'
          MountPoint: '{{ disk_preset_list[index].value.mount_point }}'
          Hostname: '{{ ansible_hostname }}'
          Instance: '{{ ansible_ec2_instance_id }}'
      register: aws_volume_info_list
      async: 1000
      poll: 0
      loop: "{{ device_name_aws_list }}"
      loop_control:
        index_var: index

    - name: Wait for volumes to be created
      async_status:
        jid: "{{ item.ansible_job_id }}"
      register: volume_create_status
      until: volume_create_status.finished
      retries: 100
      delay: 3
      loop: "{{ aws_volume_info_list.results }}"
      loop_control:
        index_var: index
```

This will asynchronously create and mount EC2 volumes against an instance (and if ansible is run against multiple instances, it will correctly do this for all 100 machines that ansible is SSH'ed to at once). If we have 10 disks to attach to a single instance, the `async` keyword here will create and attach 10 disks, and only in the second task do we wait until we have successfully created all the disks + attached them.

This looks like this:

1. VM A: Attach disk 1A, 2A, 3A, 4A, 5A
2. VM B: Attach disk 1B, 2B, 3B, 4B, 5B
3. VM C: Attach disk 1C, 2C, 3C, 4C, 5C
4. VM A: Wait for success attaching all disks
5. VM B: wait for success attaching all disks
6. VM C: wait for success attaching all disks

Basically on line 1, 2 and 3 we fire off 5 EC2 volume creation + attach requests to the EC2 API, this happens concurrently, then on line 4, 5, and 6 we concurrently wait for the async task to have completed successfully.

This means we can very quickly create disks across multiple VM's and not be stuck waiting in the following loop:

1. VM A: Attach disk 1A
2. VM A: Attach disk 2A
3. ...
4. VM A: Attach disk 5A
5. VM B: Attach disk 1B
6. VM B: Attach disk 2B
7. ...
8. VM B: Attach disk 5B

This would take much longer than firing off multiple requests to the EC2 API and only later checking/validating that they are complete.

**Comment by Fred-sun** [comment]
@rbreitendev @archoversight  Does the following way of mounting disks meet your needs?
```e.g.
  - name: Create resource group
      azure_rm_resourcegroup:
        name: "{{ resource_group }}"
        location: eastus
    - name: Create virtual network
      azure_rm_virtualnetwork:
        name: My_Virtual_Network
        address_prefixes_cidr:
          - 10.1.0.0/16
          - 172.100.0.0/16
        dns_servers:
          - 127.0.0.1
          - 127.0.0.3
        resource_group: "{{ resource_group }}"
    - name: Create subnet
      azure_rm_subnet:
        name: foobar
        virtual_network_name: My_Virtual_Network
        resource_group: "{{ resource_group }}"
        address_prefix_cidr: "10.1.0.0/24"
    - name: Create a VM with Linux OS
      azure_rm_virtualmachine:
        resource_group: "{{ resource_group }}"
        name: testvm
        vm_size: Standard_D4s_v3
        managed_disk_type: Premium_LRS
        admin_password: *************
        admin_username: *************
        ssh_public_keys:
          - path: /home/azureuser/.ssh/authorized_keys
            key_data: "ssh-rsa********************"
        image:
          offer: UbuntuServer
          publisher: Canonical
          sku: 16.04-LTS
          version: latest
    - name: attach data disk
      azure_rm_manageddisk:
        name: "disk{{ item }}"
        resource_group: "{{ resource_group }}"
        disk_size_gb: 1024
        attach_caching: read_only
        managed_by: testvm
        lun: "{{ item }}"
        zone: 1
      loop:
        - 1
        - 2
        - 3
        - 4
      async: 500
      poll: 0
      register: async_loop

**Comment by abikouo** [comment]
@archoversight I tried to simplify the way disks are created and attached. I have developed a new module (if the solution is validated we can find a way to merge into the existing module ``azure_rm_manageddisk``) into #936 

This has been tested with the following playbook:

```
- hosts: localhost
  gather_facts: no

  vars:
    test_resource_group: "aubin-manageddisk-rg"
    disk_prefix: "xxf"
    number_disk: 10

  tasks:
    - set_fact:
        virtual_machine1:
          resource_group: "{{ test_resource_group }}"
          name: "aubin-manageddisk-vm-1"
        virtual_machine2:
          resource_group: "{{ test_resource_group }}"
          name: "aubin-manageddisk-vm-2"
        virtual_machine3:
          resource_group: "{{ test_resource_group }}"
          name: "aubin-manageddisk-vm-3"
        disks_1: []
        disks_2: []
        disks_3: []


    - set_fact:
        disks_1: "{{ disks_1 + [{'disk_size_gb': 1, 'name': 'disk-' ~ vm_name ~ disk_prefix ~ item, 'resource_group': test_resource_group}] }}"
      with_sequence: start=1 end="{{ number_disk }}"
      vars:
        vm_name: "01-"
    
    - set_fact:
        disks_2: "{{ disks_1 + [{'disk_size_gb': 1, 'name': 'disk-' ~ vm_name ~ disk_prefix ~ item, 'resource_group': test_resource_group}] }}"
      with_sequence: start=1 end="{{ number_disk }}"
      vars:
        vm_name: "02-"

    - set_fact:
        disks_3: "{{ disks_1 + [{'disk_size_gb': 1, 'name': 'disk-' ~ vm_name ~ disk_prefix ~ item, 'resource_group': test_resource_group}] }}"
      with_sequence: start=1 end="{{ number_disk }}"
      vars:
        vm_name: "03-"


    - name: Manage multiple disks
      azure.azcollection.azure_rm_managedmultipledisk:
        managed_disks: "{{ item.disks }}"
        managed_by_extended:
          - "{{ item.virtual_machine }}"
      register: azure_disks
      async: 1000
      poll: 0
      with_items:
        - virtual_machine: "{{ virtual_machine1 }}"
          disks: "{{ disks_1 }}"
        - virtual_machine: "{{ virtual_machine2 }}"
          disks: "{{ disks_2 }}"
        - virtual_machine: "{{ virtual_machine3 }}"
          disks: "{{ disks_3 }}"

    - name: Wait for disks to be created and attached
      async_status:
        jid: "{{ item.ansible_job_id }}"
      register: attach_disk
      until: attach_disk.finished
      retries: 100
      delay: 5
      loop: "{{ azure_disks.results }}"
```

I had some traces to see how resources are processed, here after the detail

```
16:53:50:968689-1487411-0: Create disk aubin-manageddisk-rg/disk-01-xxf1 [Start]
16:53:51:985309-1487439-0: Create disk aubin-manageddisk-rg/disk-01-xxf1 [Start]
16:53:52:026698-1487464-0: Create disk aubin-manageddisk-rg/disk-01-xxf1 [Start]
16:53:53:010852-1487411-1: Create disk aubin-manageddisk-rg/disk-01-xxf1 [End]
16:53:53:149430-1487464-1: Create disk aubin-manageddisk-rg/disk-01-xxf1 [End]
16:53:53:347087-1487439-1: Create disk aubin-manageddisk-rg/disk-01-xxf1 [End]
16:53:55:419524-1487411-2: Create disk aubin-manageddisk-rg/disk-01-xxf2 [Start]
16:53:55:548514-1487464-2: Create disk aubin-manageddisk-rg/disk-01-xxf2 [Start]
16:53:55:748946-1487439-2: Create disk aubin-manageddisk-rg/disk-01-xxf2 [Start]
16:53:56:341010-1487411-3: Create disk aubin-manageddisk-rg/disk-01-xxf2 [End]
16:53:56:611114-1487439-3: Create disk aubin-manageddisk-rg/disk-01-xxf2 [End]
16:53:57:147548-1487464-3: Create disk aubin-manageddisk-rg/disk-01-xxf2 [End]
16:53:58:742395-1487411-4: Create disk aubin-manageddisk-rg/disk-01-xxf3 [Start]
16:53:59:014897-1487439-4: Create disk aubin-manageddisk-rg/disk-01-xxf3 [Start]
16:53:59:673374-1487464-4: Create disk aubin-manageddisk-rg/disk-01-xxf3 [Start]
16:53:59:803011-1487439-5: Create disk aubin-manageddisk-rg/disk-01-xxf3 [End]
16:54:00:115970-1487411-5: Create disk aubin-manageddisk-rg/disk-01-xxf3 [End]
16:54:00:468640-1487464-5: Create disk aubin-manageddisk-rg/disk-01-xxf3 [End]
16:54:02:234177-1487439-6: Create disk aubin-manageddisk-rg/disk-01-xxf4 [Start]
16:54:02:534208-1487411-6: Create disk aubin-manageddisk-rg/disk-01-xxf4 [Start]
16:54:02:949673-1487411-7: Create disk aubin-manageddisk-rg/disk-01-xxf4 [End]
16:54:03:097223-1487439-7: Create disk aubin-manageddisk-rg/disk-01-xxf4 [End]
16:54:03:180653-1487464-6: Create disk aubin-manageddisk-rg/disk-01-xxf5 [Start]
16:54:04:043781-1487464-7: Create disk aubin-manageddisk-rg/disk-01-xxf5 [End]
16:54:05:565497-1487411-8: Create disk aubin-manageddisk-rg/disk-01-xxf6 [Start]
16:54:05:685701-1487439-8: Create disk aubin-manageddisk-rg/disk-01-xxf6 [Start]
16:54:06:370429-1487411-9: Create disk aubin-manageddisk-rg/disk-01-xxf6 [End]
16:54:06:534037-1487439-9: Create disk aubin-manageddisk-rg/disk-01-xxf6 [End]
16:54:06:628390-1487464-8: Create disk aubin-manageddisk-rg/disk-01-xxf7 [Start]
16:54:07:430626-1487464-9: Create disk aubin-manageddisk-rg/disk-01-xxf7 [End]
16:54:08:943140-1487411-10: Create disk aubin-manageddisk-rg/disk-01-xxf8 [Start]
16:54:09:140919-1487439-10: Create disk aubin-manageddisk-rg/disk-01-xxf8 [Start]
16:54:09:923255-1487411-11: Create disk aubin-manageddisk-rg/disk-01-xxf8 [End]
16:54:10:073323-1487439-11: Create disk aubin-manageddisk-rg/disk-01-xxf8 [End]
16:54:10:143703-1487464-10: Create disk aubin-manageddisk-rg/disk-01-xxf9 [Start]
16:54:10:979598-1487464-11: Create disk aubin-manageddisk-rg/disk-01-xxf9 [End]
16:54:12:546085-1487411-12: Create disk aubin-manageddisk-rg/disk-01-xxf10 [Start]
16:54:12:737914-1487439-12: Create disk aubin-manageddisk-rg/disk-01-xxf10 [Start]
16:54:13:410268-1487464-12: Create disk aubin-manageddisk-rg/disk-01-xxf10 [Start]
16:54:13:435047-1487411-13: Create disk aubin-manageddisk-rg/disk-01-xxf10 [End]
16:54:13:679943-1487439-13: Create disk aubin-manageddisk-rg/disk-01-xxf10 [End]
16:54:13:759949-1487464-13: Create disk aubin-manageddisk-rg/disk-01-xxf10 [End]
16:54:15:971465-1487411-14: Attach disks to virtual machine aubin-manageddisk-rg/aubin-manageddisk-vm-1 [Start]
16:54:16:102590-1487439-14: Create disk aubin-manageddisk-rg/disk-02-xxf10 [Start]
16:54:16:180138-1487464-14: Create disk aubin-manageddisk-rg/disk-03-xxf10 [Start]
16:54:16:994936-1487464-15: Create disk aubin-manageddisk-rg/disk-03-xxf10 [End]
16:54:17:015767-1487439-15: Create disk aubin-manageddisk-rg/disk-02-xxf10 [End]
16:54:19:509482-1487464-16: Attach disks to virtual machine aubin-manageddisk-rg/aubin-manageddisk-vm-3 [Start]
16:54:19:543706-1487439-16: Attach disks to virtual machine aubin-manageddisk-rg/aubin-manageddisk-vm-2 [Start]
```

Please provide feedback

**Comment by Fred-sun** [comment]
@abikouo Something like this?

```e.g.
    - name: Create resource group
      azure_rm_resourcegroup:
        name: "{{ resource_group }}"
        location: eastus

    - name: Create virtual network
      azure_rm_virtualnetwork:
        name: My_Virtual_Network
        address_prefixes_cidr:
          - 10.1.0.0/16
          - 172.100.0.0/16
        dns_servers:
          - 127.0.0.1
          - 127.0.0.3
        resource_group: "{{ resource_group }}"

    - name: Create subnet
      azure_rm_subnet:
        name: foobar
        virtual_network_name: My_Virtual_Network
        resource_group: "{{ resource_group }}"
        address_prefix_cidr: "10.1.0.0/24"

    - name: Create a VM with Linux OS  --- first
      azure_rm_virtualmachine:
        resource_group: "{{ resource_group }}"
        name: testvm01
        vm_size: Standard_D4s_v3
        managed_disk_type: Premium_LRS
        admin_password: Password@0329
        admin_username: azureuser
        ssh_public_keys:
          - path: /home/azureuser/.ssh/authorized_keys
            key_data: "ssh-rsa xxxxx"
        image:
          offer: UbuntuServer
          publisher: Canonical
          sku: 16.04-LTS
          version: latest

    - name: Create a VM with Linux OS --- secondary
      azure_rm_virtualmachine:
        resource_group: "{{ resource_group }}"
        name: testvm02
        vm_size: Standard_D4s_v3
        managed_disk_type: Premium_LRS
        admin_password: Password@0329
        admin_username: azureuser
        ssh_public_keys:
          - path: /home/azureuser/.ssh/authorized_keys
            key_data: "ssh-rsa xxxxx"
        image:
          offer: UbuntuServer
          publisher: Canonical
          sku: 16.04-LTS
          version: latest

    - name: Create a VM with Linux OS --- third
      azure_rm_virtualmachine:
        resource_group: "{{ resource_group }}"
        name: testvm03
        vm_size: Standard_D4s_v3
        managed_disk_type: Premium_LRS
        admin_password: Password@0329
        admin_username: azureuser
        ssh_public_keys:
          - path: /home/azureuser/.ssh/authorized_keys
            key_data: "ssh-rsa xxxxx"
        image:
          offer: UbuntuServer
          publisher: Canonical
          sku: 16.04-LTS
          version: latest

    - set_fact:
        virtual_machine1:
          vm_name: testvm01
        virtual_machine2:
          vm_name: testvm02
        virtual_machine3:
          vm_name: testvm03

    - set_fact:
        disk1:
          disk_name: testdisk01
        disk2:
          disk_name: testdisk02
        disk3:
          disk_name: testdisk03

    - name: attach data disk
      azure_rm_manageddisk:
        name: "{{ item.disk.disk_name }}"
        resource_group: "{{ resource_group }}"
        disk_size_gb: 1024
        attach_caching: read_only
        managed_by: "{{ item.virtual_machine.vm_name }}"
        zone: 1
      with_items:
        - virtual_machine: "{{ virtual_machine1 }}"
          disk: "{{ disk1 }}"
        - virtual_machine: "{{ virtual_machine2 }}"
          disk: "{{ disk2 }}"
        - virtual_machine: "{{ virtual_machine3 }}"
          disk: "{{ disk3 }}"
      async: 1000
      poll: 0
      register: async_loop

    - name: wait
      async_status:
        jid: "{{item.ansible_job_id}}"
      retries: 100
      delay: 5
      loop: "{{async_loop.results}}"
      register: async_loop_jobs
      until: async_loop_jobs.finished

**Comment by ohthehugemanatee** [comment]
@rbreitendev @archoversight  

I'm sorry, can you please confirm how many layers of async you want? The text descriptions just aren't clear enough for me. Is it:


**Option A**:
![image](https://user-images.githubusercontent.com/907257/193047625-9b5c7700-9add-4717-bb55-630bbf56ba2b.png)

or **Option B**:
![image](https://user-images.githubusercontent.com/907257/193047706-558f8726-4da3-4212-92e8-f57e5d636d96.png)

(to clarify, every disk in these diagrams is an individual disk. No disks are shared between VMs. I should have named them differently in the diagram, sorry!)

It looks like PR #936  solves for idempotency and Option A... but that's not the whole solution, right?

**Comment by archoversight** [comment]
@ohthehugemanatee creation of the VM's is done using terraform, and is out of scope.

We care about creation of disks and attaching them being async, so that when add a lot of disks it happens quickly.

 That seems to be taken care of in https://github.com/ansible-collections/azure/pull/936 partially, in that its creating the disks one after the other, then attaching all of them at the end (if I am reading it correctly).
 
 So it would:

- create disk 1
- create disk 2
- create disk 3
- Attach disk 1, 2, 3 to VM

This still reduces contention due to concurrent modifications to the VM object which would fail.

Having async on the create disk step (so that multiple create disks could be fired off at once) would be great to reduce the time even further.

**Comment by abikouo** [comment]
@archoversight I pushed a new commit to create multiple disks at the same time, the azure API is working asynchronously, basically with this new version we submit the creation for all disks then later we poll for the result.
Please have a look and make some tests on your side.
